File: frontend/src/agent/knowledge.ts
import { apiUrl, getJson, postJson } from '../lib/api'
import { authHeader } from '../auth/token'

export interface SourceRow {
  source: string
  chunks: number
  kind?: 'image' | 'pdf' | 'document' | 'link' | 'text'
  content_type?: string
  url?: string
}

// Absolute, public URL of a source's stored original file. Absolute so an
// embedded document viewer (Google Docs Viewer) can fetch it.
export function knowledgeFileUrl(source: string): string {
  const u = apiUrl(`/api/knowledge/file/${encodeURIComponent(source)}`)
  return u.startsWith('http') ? u : `${window.location.origin}${u}`
}

export interface IngestResult {
  ok: boolean
  source: string
  chunks: number
}

export async function listSources(): Promise<SourceRow[]> {
  return getJson<SourceRow[]>('/api/knowledge')
}

// The indexed plain text of a source (for previewing a note that has no file).
export async function getSourceContent(source: string): Promise<{ source: string; text: string }> {
  return getJson<{ source: string; text: string }>(
    `/api/knowledge/content/${encodeURIComponent(source)}`,
  )
}

export async function ingestUrl(url: string, source?: string): Promise<IngestResult> {
  return postJson<IngestResult>('/api/knowledge/url', { url, source: source ?? null })
}

export async function ingestText(text: string, source: string): Promise<IngestResult> {
  return postJson<IngestResult>('/api/knowledge/text', { text, source })
}

async function postForm<T>(path: string, form: FormData): Promise<T> {
  const res = await fetch(apiUrl(path), {
    method: 'POST',
    headers: { ...authHeader() },
    body: form,
  })
  if (!res.ok) {
    const text = await res.text().catch(() => '')
    throw new Error(`POST ${path} failed: ${res.status} ${res.statusText} ${text}`)
  }
  return res.json() as Promise<T>
}

export async function ingestPdf(file: File, source?: string): Promise<IngestResult> {
  const form = new FormData()
  form.append('file', file)
  if (source) form.append('source', source)
  return postForm<IngestResult>('/api/knowledge/pdf', form)
}

export async function ingestImage(file: File, source?: string): Promise<IngestResult> {
  const form = new FormData()
  form.append('file', file)
  if (source) form.append('source', source)
  return postForm<IngestResult>('/api/knowledge/image', form)
}

// Word / PowerPoint / Excel / HTML / EPUB / CSV, converted to Markdown on the
// backend (markitdown) and embedded with Gemini like any other source.
export async function ingestDocument(file: File, source?: string): Promise<IngestResult> {
  const form = new FormData()
  form.append('file', file)
  if (source) form.append('source', source)
  return postForm<IngestResult>('/api/knowledge/document', form)
}

export async function deleteSource(source: string): Promise<void> {
  const res = await fetch(apiUrl(`/api/knowledge/${encodeURIComponent(source)}`), {
    method: 'DELETE',
    headers: { ...authHeader() },
  })
  if (!res.ok) throw new Error(`delete failed: ${res.status}`)
}

File: backend/app/routes/knowledge.py
"""Knowledge base (RAG) admin endpoints.

Lets the app ingest documentation into the `knowledge_chunks` collection so the
agent's `search_docs` tool can cite it. Supports PDF, web URL, raw text, and
image sources - all embedded with Gemini (768 dims)."""

from __future__ import annotations

from typing import Annotated

import httpx
from fastapi import APIRouter, File, Form, HTTPException, Response, UploadFile
from pydantic import BaseModel

from ..services import knowledge as kb
from ..services import knowledge_files as kbf

router = APIRouter()

_MAX_UPLOAD_BYTES = 50 * 1024 * 1024


class SourceRow(BaseModel):
    source: str
    chunks: int
    # How the UI should show it: image | pdf | document | link | text.
    kind: str = "text"
    content_type: str | None = None
    url: str | None = None


class IngestResult(BaseModel):
    ok: bool
    source: str
    chunks: int


class SourceContent(BaseModel):
    source: str
    text: str


class UrlIngestRequest(BaseModel):
    url: str
    source: str | None = None


class TextIngestRequest(BaseModel):
    text: str
    source: str


@router.get("", response_model=list[SourceRow])
async def list_sources() -> list[SourceRow]:
    rows = await kb.list_sources()
    return [SourceRow(**row) for row in rows]


@router.post("/url", response_model=IngestResult)
async def ingest_url(payload: UrlIngestRequest) -> IngestResult:
    try:
        count = await kb.index_url(payload.url, source=payload.source)
    except httpx.HTTPError as exc:
        raise HTTPException(400, f"could not fetch url: {exc}") from exc
    label = payload.source or payload.url
    return IngestResult(ok=count > 0, source=label, chunks=count)


@router.post("/text", response_model=IngestResult)
async def ingest_text(payload: TextIngestRequest) -> IngestResult:
    count = await kb.index_plain_text(payload.text, source=payload.source)
    return IngestResult(ok=count > 0, source=payload.source, chunks=count)


@router.post("/pdf", response_model=IngestResult)
async def ingest_pdf(
    file: Annotated[UploadFile, File()],
    source: Annotated[str | None, Form()] = None,
) -> IngestResult:
    data = await file.read()
    if len(data) > _MAX_UPLOAD_BYTES:
        raise HTTPException(413, "file too large")
    label = source or (file.filename or "uploaded.pdf")
    try:
        count = await kb.index_pdf_bytes(data, source=label)
    except Exception as exc:
        raise HTTPException(400, f"could not read pdf: {exc}") from exc
    return IngestResult(ok=count > 0, source=label, chunks=count)


@router.post("/document", response_model=IngestResult)
async def ingest_document(
    file: Annotated[UploadFile, File()],
    source: Annotated[str | None, Form()] = None,
) -> IngestResult:
    """Ingest a Word/PowerPoint/Excel/HTML/EPUB/CSV document (converted to
    Markdown via markitdown). Use /pdf for PDFs and /image for images."""
    from ..services.doc_extract import is_supported

    data = await file.read()
    if len(data) > _MAX_UPLOAD_BYTES:
        raise HTTPException(413, "file too large")
    name = file.filename or "document"
    if not is_supported(name):
        raise HTTPException(400, f"unsupported document type: {name}. Use /pdf or /image for those.")
    label = source or name
    try:
        count = await kb.index_document_bytes(data, filename=name, source=label)
    except Exception as exc:
        raise HTTPException(400, f"could not read document: {exc}") from exc
    return IngestResult(ok=count > 0, source=label, chunks=count)


@router.post("/image", response_model=IngestResult)
async def ingest_image(
    file: Annotated[UploadFile, File()],
    source: Annotated[str | None, Form()] = None,
) -> IngestResult:
    data = await file.read()
    if len(data) > _MAX_UPLOAD_BYTES:
        raise HTTPException(413, "file too large")
    label = source or (file.filename or "image")
    mime = file.content_type or "image/png"
    count = await kb.index_image(data, mime_type=mime, source=label)
    return IngestResult(ok=count > 0, source=label, chunks=count)


@router.get("/content/{source:path}", response_model=SourceContent)
async def get_content(source: str) -> SourceContent:
    """The indexed plain text of a source, so the UI can preview a note."""
    return SourceContent(source=source, text=await kb.get_source_text(source))


@router.get("/file/{source:path}")
async def get_file(source: str) -> Response:
    """Stream the original file stored for a source (so the UI can preview a PDF,
    image, or document). Public, so an embedded document viewer can fetch it."""
    result = await kbf.get_file(source)
    if result is None:
        raise HTTPException(404, "no file stored for this source")
    content_type, data = result
    return Response(content=data, media_type=content_type)


@router.delete("/{source:path}", response_model=dict)
async def delete_source(source: str) -> dict:
    deleted = await kb.delete_source(source)
    return {"ok": True, "deleted": deleted}
